File: src/ee/vungle/VungleBridge.cpp
#include "ee/vungle/VungleBridge.hpp"

#include <ee/nlohmann/json.hpp>

#include <ee/ads/internal/GuardedRewardedAd.hpp>
#include <ee/ads/internal/IAsyncHelper.hpp>
#include <ee/ads/internal/MediationManager.hpp>
#include <ee/core/Logger.hpp>
#include <ee/core/PluginManager.hpp>
#include <ee/core/Utils.hpp>
#include <ee/core/internal/MessageBridge.hpp>

#include "ee/vungle/private/VungleRewardedAd.hpp"

namespace ee {
namespace vungle {
using Self = Bridge;

namespace {
// clang-format off
const std::string kPrefix    = "VungleBridge";

const auto kInitialize     = kPrefix + "Initialize";

const auto kHasRewardedAd  = kPrefix + "HasRewardedAd";
const auto kLoadRewardedAd = kPrefix + "LoadRewardedAd";
const auto kShowRewardedAd = kPrefix + "ShowRewardedAd";

const auto kOnLoaded       = kPrefix + "OnLoaded";
const auto kOnFailedToLoad = kPrefix + "OnFailedToLoad";
const auto kOnFailedToShow = kPrefix + "OnFailedToShow";
const auto kOnClicked      = kPrefix + "OnClicked";
const auto kOnClosed       = kPrefix + "OnClosed";
// clang-format on
} // namespace

Self::Bridge()
    : Self(Logger::getSystemLogger()) {}

Self::Bridge(const Logger& logger)
    : bridge_(MessageBridge::getInstance())
    , logger_(logger) {
    logger_.debug(__PRETTY_FUNCTION__);
    PluginManager::addPlugin(Plugin::Vungle);

    auto&& mediation = ads::MediationManager::getInstance();
    rewardedAdDisplayer_ = mediation.getRewardedAdDisplayer();

    bridge_.registerHandler(
        [this](const std::string& message) {
            auto json = nlohmann::json::parse(message);
            onLoaded(json["ad_id"]);
            return "";
        },
        kOnLoaded);
    bridge_.registerHandler(
        [this](const std::string& message) {
            auto json = nlohmann::json::parse(message);
            onFailedToLoad(json["ad_id"], json["message"]);
            return "";
        },
        kOnFailedToLoad);
    bridge_.registerHandler(
        [this](const std::string& message) {
            auto json = nlohmann::json::parse(message);
            onFailedToShow(json["ad_id"], json["message"]);
            return "";
        },
        kOnFailedToShow);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onClicked(message);
            return "";
        },
        kOnClicked);
    bridge_.registerHandler(
        [this](const std::string& message) {
            auto json = nlohmann::json::parse(message);
            onClosed(json["ad_id"], json["rewarded"]);
            return "";
        },
        kOnClosed);
}

Self::~Bridge() = default;

void Self::destroy() {
    logger_.debug(__PRETTY_FUNCTION__);
    bridge_.deregisterHandler(kOnLoaded);
    bridge_.deregisterHandler(kOnFailedToLoad);
    bridge_.deregisterHandler(kOnFailedToShow);
    bridge_.deregisterHandler(kOnClosed);
    PluginManager::removePlugin(Plugin::Vungle);
}

void Self::initialize(const std::string& appId) {
    logger_.debug("%s: appId = %s", __PRETTY_FUNCTION__, appId.c_str());
    nlohmann::json json;
    json["appId"] = appId;
    bridge_.call(kInitialize, json.dump());
}

std::shared_ptr<IRewardedAd> Self::createRewardedAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        return iter->second.strong;
    }
    auto weak = new RewardedAd(logger_, rewardedAdDisplayer_, this, adId);
    auto strong = std::make_shared<ads::GuardedRewardedAd>(
        std::shared_ptr<RewardedAd>(weak));
    rewardedAds_.try_emplace(adId, strong, weak);
    return strong;
}

bool Self::destroyRewardedAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    auto iter = rewardedAds_.find(adId);
    if (iter == rewardedAds_.cend()) {
        return false;
    }
    rewardedAds_.erase(iter);
    return true;
}

bool Self::hasRewardedAd(const std::string& adId) const {
    auto result = bridge_.call(kHasRewardedAd, adId);
    return core::toBool(result);
}

void Self::loadRewardedAd(const std::string& adId) const {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    bridge_.call(kLoadRewardedAd, adId);
}

void Self::showRewardedAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    bridge_.call(kShowRewardedAd, adId);
}

void Self::onLoaded(const std::string& adId) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onLoaded();
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onFailedToLoad(const std::string& adId, const std::string& message) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onFailedToLoad(message);
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onFailedToShow(const std::string& adId, const std::string& message) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onFailedToShow(message);
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onClicked(const std::string& adId) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onClicked();
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onClosed(const std::string& adId, bool rewarded) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onClosed(rewarded);
    } else {
        // Mediation.
        assert(false);
        onMediationAdClosed(adId, rewarded);
    }
}

void Self::onMediationAdClosed(const std::string& adId, bool rewarded) {
    logger_.debug("%s: %s", __PRETTY_FUNCTION__,
                  core::toString(rewarded).c_str());
    if (rewardedAdDisplayer_->isProcessing()) {
        rewardedAdDisplayer_->resolve(rewarded ? IRewardedAdResult::Completed
                                               : IRewardedAdResult::Canceled);
        return;
    }
    assert(false);
}
} // namespace vungle
} // namespace ee

File: proj.android_studio/ee-x-vungle/src/main/java/com/ee/VungleBridge.kt
package com.ee

import android.app.Activity
import android.content.Context
import androidx.annotation.AnyThread
import com.ee.internal.deserialize
import com.ee.internal.serialize
import com.vungle.warren.AdConfig
import com.vungle.warren.InitCallback
import com.vungle.warren.LoadAdCallback
import com.vungle.warren.PlayAdCallback
import com.vungle.warren.Vungle
import com.vungle.warren.error.VungleException
import kotlinx.serialization.ImplicitReflectionSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.UnstableDefault
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by Pham Xuan Han on 17/05/17.
 */
@ImplicitReflectionSerializer
@UnstableDefault
class VungleBridge(
    private val _bridge: IMessageBridge,
    private val _context: Context,
    private var _activity: Activity?) : IPlugin {
    companion object {
        private val _logger = Logger(VungleBridge::class.java.name)

        private const val kPrefix = "VungleBridge"
        private const val kInitialize = "${kPrefix}Initialize"
        private const val kHasRewardedAd = "${kPrefix}HasRewardedAd"
        private const val kLoadRewardedAd = "${kPrefix}LoadRewardedAd"
        private const val kShowRewardedAd = "${kPrefix}ShowRewardedAd"
        private const val kOnLoaded = "${kPrefix}OnLoaded"
        private const val kOnFailedToLoad = "${kPrefix}OnFailedToLoad"
        private const val kOnFailedToShow = "${kPrefix}OnFailedToShow"
        private const val kOnClicked = "${kPrefix}OnClicked"
        private const val kOnClosed = "${kPrefix}OnClosed"
    }

    private var _initializing = false
    private var _initialized = false
    private var _rewarded = false
    private val _loadedAdIds: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap())

    init {
        _logger.debug("constructor begin: context = $_context activity = $_activity")
        registerHandlers()
        _logger.debug("constructor end.")
    }

    override fun onCreate(activity: Activity) {}
    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {}
    override fun onPause() {}
    override fun onDestroy() {}

    override fun destroy() {
        deregisterHandlers()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(kInitialize) { message ->
            @Serializable
            class Request(
                val application_id: String
            )

            val request = deserialize<Request>(message)
            initialize(request.application_id)
            ""
        }
        _bridge.registerHandler(kHasRewardedAd) { message ->
            Utils.toString(hasRewardedAd(message))
        }
        _bridge.registerHandler(kLoadRewardedAd) { message ->
            loadRewardedAd(message)
            ""
        }
        _bridge.registerHandler(kShowRewardedAd) { message ->
            showRewardedAd(message)
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kHasRewardedAd)
        _bridge.deregisterHandler(kShowRewardedAd)
        _bridge.deregisterHandler(kLoadRewardedAd)
    }

    @AnyThread
    fun initialize(applicationId: String) {
        Thread.runOnMainThread(Runnable {
            if (_initializing) {
                return@Runnable
            }
            if (_initialized) {
                return@Runnable
            }
            _initializing = true

            Vungle.init(applicationId, _context, object : InitCallback {
                override fun onSuccess() {
                    _logger.info("${VungleBridge::initialize.name}: ${this::onSuccess.name}")
                    _initializing = false
                    _initialized = true
                }

                override fun onError(throwable: VungleException) {
                    _logger.info("${VungleBridge::initialize.name}: ${this::onError.name}: message = ${throwable.localizedMessage}")
                    _initializing = false
                }

                override fun onAutoCacheAdAvailable(adId: String) {}
            })
        })
    }

    @AnyThread
    private fun hasRewardedAd(adId: String): Boolean {
        return _loadedAdIds.contains(adId)
    }

    @AnyThread
    private fun loadRewardedAd(adId: String) {
        Thread.runOnMainThread(Runnable {
            Vungle.loadAd(adId, object : LoadAdCallback {
                override fun onAdLoad(adId: String) {
                    _logger.info("${VungleBridge::loadRewardedAd.name}: ${this::onAdLoad.name}: $adId")
                    _loadedAdIds.add(adId)

                    @Serializable
                    @Suppress("unused")
                    class Response(
                        val ad_id: String
                    )

                    val response = Response(adId)
                    _bridge.callCpp(kOnLoaded, response.serialize())
                }

                override fun onError(adId: String, exception: VungleException) {
                    _logger.info("${VungleBridge::loadRewardedAd.name}: ${this::onError.name}: $adId reason: ${exception.localizedMessage}")
                    @Serializable
                    @Suppress("unused")
                    class Response(
                        val ad_id: String,
                        val message: String
                    )

                    val response = Response(adId, exception.localizedMessage ?: "")
                    _bridge.callCpp(kOnFailedToLoad, response.serialize())
                }
            })
        })
    }

    @AnyThread
    private fun showRewardedAd(adId: String) {
        Thread.runOnMainThread(Runnable {
            _rewarded = false
            Vungle.playAd(adId, AdConfig(), object : PlayAdCallback {
                override fun onError(adId: String, exception: VungleException) {
                    _logger.info("${VungleBridge::showRewardedAd.name}: ${this::onError.name}: $adId message = ${exception.localizedMessage}")
                    @Serializable
                    @Suppress("unused")
                    class Response(
                        val ad_id: String,
                        val message: String
                    )

                    val response = Response(adId, exception.localizedMessage ?: "")
                    _bridge.callCpp(kOnFailedToShow, response.serialize())
                }

                override fun onAdStart(adId: String) {
                    _logger.info("${VungleBridge::showRewardedAd.name}: ${this::onAdStart.name}: $adId")
                }

                override fun onAdClick(adId: String) {
                    _logger.info("${VungleBridge::showRewardedAd.name}: ${this::onAdClick.name}: $adId")
                    _bridge.callCpp(kOnClicked, adId)
                }

                override fun onAdLeftApplication(adId: String) {
                    _logger.info("${VungleBridge::showRewardedAd.name}: ${this::onAdLeftApplication.name}: $adId")
                }

                override fun onAdRewarded(adId: String) {
                    _logger.info("${VungleBridge::showRewardedAd.name}: ${this::onAdRewarded.name}: $adId")
                    _rewarded = true
                }

                override fun onAdEnd(adId: String) {
                    _logger.info("${VungleBridge::showRewardedAd.name}: onAdEnd: $adId")
                    @Serializable
                    @Suppress("unused")
                    class Response(
                        val ad_id: String,
                        val rewarded: Boolean
                    )

                    _loadedAdIds.remove(adId)
                    val response = Response(adId, _rewarded)
                    _bridge.callCpp(kOnClosed, response.serialize())
                }

                override fun onAdEnd(adId: String, completed: Boolean, isCTAClicked: Boolean) {
                    // Deprecated.
                }
            })
        })
    }
}